lib/mayaUsd/render/MaterialXGenOgsXml/GlslOcioNodeImpl.cpp

```diff
@@ -0,0 +1,369 @@
+    if (!theRenderer) {
+        return {};
+    }
+
+    MHWRender::MFragmentManager* fragmentManager = theRenderer->getFragmentManager();
+    if (!fragmentManager) {
+        return {};
+    }
+
+    MString fragText;
+    if (!fragmentManager->getFragmentXML(fragName.c_str(), fragText)) {
+        return {};
+    }
+
+    pugi::xml_document     doc;
+    pugi::xml_parse_result result = doc.load_string(fragText.asChar());
+
+    if (!result) {
+        MString errorMsg = "XML error parsing fragment for ";
+        errorMsg += fragName.c_str();
+        errorMsg += " at character ";
+        errorMsg += static_cast<int>(result.offset);
+        errorMsg += ": ";
+        errorMsg += result.description();
+        MGlobal::displayError(errorMsg);
+        return {};
+    }
+
+    // Validate that the fragment structure is 100% as expected. We need the properties, values,
+    // outputs, and GLSL source to proceed:
+    auto fragment = doc.child(TAG_FRAGMENT);
+    if (!fragment) {
+        return {};
+    }
+
+    auto properties = fragment.child(TAG_PROPERTIES);
+    if (!properties) {
+        return {};
+    }
+
+    auto values = fragment.child(TAG_VALUES);
+    if (!values) {
+        return {};
+    }
+
+    auto outputs = fragment.child(TAG_OUTPUTS);
+    if (!outputs) {
+        return {};
+    }
+
+    auto implementations = fragment.child(TAG_IMPLEMENTATION);
+    if (!implementations) {
+        return {};
+    }
+    bool hasGLSL = false;
+    for (auto&& implementation : implementations.children()) {
+        auto language = implementation.attribute(ATTR_LANGUAGE);
+        if (std::strncmp(language.as_string(), OCIO_GLSL, OCIO_GLSL_LEN) == 0) {
+            hasGLSL = true;
+            break;
+        }
+    }
+    if (!hasGLSL) {
+        return {};
+    }
+
+    // We now have all the data we need. Preserve the info and add our new OCIO NodeDef in the
+    // library.
+    addOCIONodeDef(mtlxLibrary, doc, OCIO_COLOR3);
+    addOCIONodeDef(mtlxLibrary, doc, OCIO_COLOR4);
+    knownOCIOFragments.emplace(fragName, std::move(doc));
+
+    return getUntypedNodeDefName(fragName);
+}
+
+const std::vector<std::string>& GlslOcioNodeImpl::getOCIOImplementations()
+{
+    return knownOCIOImplementations;
+}
+
+ShaderNodeImplPtr GlslOcioNodeImpl::create() { return std::make_shared<GlslOcioNodeImpl>(); }
+
+void GlslOcioNodeImpl::createVariables(const ShaderNode& node, GenContext& context, Shader& shader)
+    const
+{
+    const auto& ocioDoc = docFromImplementationName(getName());
+    if (ocioDoc.empty()) {
+        return;
+    }
+
+    // Need to create texture inputs if required:
+    for (auto&& property : ocioDoc.child(TAG_FRAGMENT).child(TAG_PROPERTIES).children()) {
+        if (std::strncmp(property.name(), TAG_TEXTURE2, TAG_TEXTURE2_LEN) == 0) {
+            std::string samplerName = property.attribute(ATTR_NAME).as_string();
+            samplerName += "Sampler";
```

Comment:
Preserving the name to make sure Maya can load the LUTs for the more complex color management shaders.